### PR TITLE
support for compute shared image gallery & support for removing any nic (primary,  secondary) from VM

### PR DESCRIPTION
**Pull request by anuchandy**
fix for https://github.com/Azure/azure-libraries-for-net/issues/355 as well

### CODE REVIEW

File: src/ResourceManagement/Compute/GalleryImageImpl.cs [hunk 1]
```diff
@@ -0,0 +1,598 @@
...
+            }
+        }
+
+
+        ///GENMHASH:5AD91481A0966B059A478CD4E9DD9466:23EC4B61BB9F1DFE8F251827F8E66FF4
+        protected override async Task<Models.GalleryImageInner> GetInnerAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            var client = this.computeManager.Inner.GalleryImages;
+            return await client.GetAsync(this.resourceGroupName, this.galleryName, this.galleryImageName);
+        }
```
Comment (hovsepm) [inline src/ResourceManagement/Compute/GalleryImageImpl.cs]: cancellationToken

File: src/ResourceManagement/Compute/GalleryImageImpl.cs [hunk 2]
```diff
@@ -0,0 +1,598 @@
...
+
+        ///GENMHASH:0202A00A1DCF248D2647DBDBEF2CA865:3EE578C74AE5F7B41F5C490A1DC7D638
+        public override async Task<Microsoft.Azure.Management.Compute.Fluent.IGalleryImage> CreateResourceAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            var client = this.computeManager.Inner.GalleryImages;
+            var inner = await client.CreateOrUpdateAsync(this.resourceGroupName, this.galleryName, this.galleryImageName, this.Inner);
+            this.SetInner(inner);
```
Comment (hovsepm) [inline src/ResourceManagement/Compute/GalleryImageImpl.cs]: cancellationToken

File: src/ResourceManagement/Compute/GalleryImageImpl.cs [hunk 3]
```diff
@@ -0,0 +1,598 @@
...
+        public IGalleryImageVersion GetVersion(string versionName)
+        {
+            return this.computeManager.GalleryImageVersions.GetByGalleryImage(this.resourceGroupName, this.galleryName, this.galleryImageName, versionName);
+        }
+
+        ///GENMHASH:7F002133997D7216F18BB8D21CF034D2:BE28E3F7E873613CF430C710146DC5B3
+        public async Task<Microsoft.Azure.Management.Compute.Fluent.IGalleryImageVersion> GetVersionAsync(string versionName, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            return await this.computeManager.GalleryImageVersions.GetByGalleryImageAsync(this.resourceGroupName, this.galleryName, this.galleryImageName, versionName);
+        }
```
Comment (hovsepm) [inline src/ResourceManagement/Compute/GalleryImageImpl.cs]: cancellationToken

File: src/ResourceManagement/Compute/GalleryImageImpl.cs [hunk 4]
```diff
@@ -0,0 +1,598 @@
...
+        public IEnumerable<Microsoft.Azure.Management.Compute.Fluent.IGalleryImageVersion> ListVersions()
+        {
+            return this.computeManager.GalleryImageVersions.ListByGalleryImage(this.resourceGroupName, this.galleryName, this.galleryImageName);
+        }
+
+        ///GENMHASH:D4A7A02E673639C444C53A8D52EFA5E3:6BDB02D858015991B4568658D3E9ADC2
+        public async Task<IPagedCollection<Microsoft.Azure.Management.Compute.Fluent.IGalleryImageVersion>> ListVersionsAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            return await this.computeManager.GalleryImageVersions.ListByGalleryImageAsync(this.resourceGroupName, this.galleryName, this.galleryImageName);
+        }
```
Comment (hovsepm) [inline src/ResourceManagement/Compute/GalleryImageImpl.cs]: cancellationToken

File: src/ResourceManagement/Compute/GalleryImagesImpl.cs [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        {
+            Extensions.Synchronize(() => this.DeleteByGalleryAsync(resourceGroupName, galleryName, galleryImageName));
+        }
+
+        ///GENMHASH:F3D774EE2FA1E468815997A41FC8FE39:59C3BBD5140129084DF92FFE9968F576
+        public async Task DeleteByGalleryAsync(string resourceGroupName, string galleryName, string galleryImageName, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            var client = this.Inner;
+            await client.DeleteAsync(resourceGroupName, galleryName, galleryImageName);
+        }
```
Comment (hovsepm) [inline src/ResourceManagement/Compute/GalleryImagesImpl.cs]: cancellationToken

File: src/ResourceManagement/Compute/GalleryImagesImpl.cs [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        public IGalleryImage GetByGallery(string resourceGroupName, string galleryName, string galleryImageName)
+        {
+            return Extensions.Synchronize(() => this.GetByGalleryAsync(resourceGroupName, galleryName, galleryImageName));
+        }
+
+        ///GENMHASH:A73D04375D3F249B3045C9CEDC2141F3:1648608B9C06972043D4D7D4160CE0CA
+        public async Task<Microsoft.Azure.Management.Compute.Fluent.IGalleryImage> GetByGalleryAsync(string resourceGroupName, string galleryName, string galleryImageName, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            var inner = await this.Inner.GetAsync(resourceGroupName, galleryName, galleryImageName);
+            return WrapModel(inner);
```
Comment (hovsepm) [inline src/ResourceManagement/Compute/GalleryImagesImpl.cs]: cancellationToken

File: src/ResourceManagement/Compute/GalleryImageVersionImpl.cs [hunk 7]
```diff
@@ -0,0 +1,435 @@
...
+            this.galleryImageVersionName = GetValueFromIdByName(inner.Id, "versions");
+        }
+
+
+        ///GENMHASH:5AD91481A0966B059A478CD4E9DD9466:0E64D352F7F18B753618D9C7E943FF2F
+        protected override async Task<Models.GalleryImageVersionInner> GetInnerAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            var client = this.computeManager.Inner.GalleryImageVersions;
+            return await client.GetAsync(this.resourceGroupName, this.galleryName, this.galleryImageName, this.galleryImageVersionName);
+        }
```
Comment (hovsepm) [inline src/ResourceManagement/Compute/GalleryImageVersionImpl.cs]: cancellationToken

File: src/ResourceManagement/Compute/GalleryImageVersionImpl.cs [hunk 8]
```diff
@@ -0,0 +1,435 @@
...
+                }
+            }
+            return regions;
+        }
+
+        ///GENMHASH:0202A00A1DCF248D2647DBDBEF2CA865:1441758C6DA19434093428FE9E67FD01
+        public override async Task<Microsoft.Azure.Management.Compute.Fluent.IGalleryImageVersion> CreateResourceAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            var inner = await this.computeManager.Inner.GalleryImageVersions.CreateOrUpdateAsync(this.resourceGroupName, this.galleryName, this.galleryImageName, this.galleryImageVersionName, this.Inner);
+            this.SetInner(inner);
```
Comment (hovsepm) [inline src/ResourceManagement/Compute/GalleryImageVersionImpl.cs]: cancellationToken

File: src/ResourceManagement/Compute/GalleryImageVersionsImpl.cs [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+        public IGalleryImageVersion GetByGalleryImage(string resourceGroupName, string galleryName, string galleryImageName, string galleryImageVersionName)
+        {
+            return Extensions.Synchronize(() => this.GetByGalleryImageAsync(resourceGroupName, galleryName, galleryImageName, galleryImageVersionName));
+        }
+
+        ///GENMHASH:86DA5629FEB54C10E005F23B4AC06853:1CB6596D9B3C89EB31276EB7A45C1C1C
+        public async Task<Microsoft.Azure.Management.Compute.Fluent.IGalleryImageVersion> GetByGalleryImageAsync(string resourceGroupName, string galleryName, string galleryImageName, string galleryImageVersionName, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            var inner = await this.Inner.GetAsync(resourceGroupName, galleryName, galleryImageName, galleryImageVersionName);
+            return this.WrapModel(inner);
```
Comment (hovsepm) [inline src/ResourceManagement/Compute/GalleryImageVersionsImpl.cs]: cancellationToken

File: src/ResourceManagement/Compute/GalleryImageVersionsImpl.cs [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+                .AsContinuousCollection(nextLink => Extensions.Synchronize(() => this.Inner.ListByGalleryImageNextAsync(nextLink)))
+                .Select(inner => WrapModel(inner));
+        }
+
+        ///GENMHASH:6850F4FC22F8DD108AC5AAC3F880BB75:ECA8E93D1C1A32FE048EEA5B2C38138D
+        public async Task<IPagedCollection<Microsoft.Azure.Management.Compute.Fluent.IGalleryImageVersion>> ListByGalleryImageAsync(string resourceGroupName, string galleryName, string galleryImageName, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            return await PagedCollection<IGalleryImageVersion, GalleryImageVersionInner>.LoadPage(cancel => this.Inner.ListByGalleryImageAsync(resourceGroupName, galleryName, galleryImageName, cancel),
+                  (nextLink, cancel) => this.Inner.ListByGalleryImageNextAsync(nextLink),
+                  WrapModel,
```
Comment (hovsepm) [inline src/ResourceManagement/Compute/GalleryImageVersionsImpl.cs]: cancellationToken

File: src/ResourceManagement/Compute/GalleryImpl.cs [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+            var galleryInner = await Manager.Inner.Galleries.CreateOrUpdateAsync(ResourceGroupName, Name, Inner, cancellationToken);
+            SetInner(galleryInner);
+            return this;
+        }
+
+        ///GENMHASH:53C5326A96805851579A861BAA315694:20E793DA2931A94A780043403389CAE6
+        public async Task<Microsoft.Azure.Management.Compute.Fluent.IGalleryImage> GetImageAsync(string imageName, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            return await this.Manager.GalleryImages.GetByGalleryAsync(this.ResourceGroupName, this.Name, imageName);
+        }
```
Comment (hovsepm) [inline src/ResourceManagement/Compute/GalleryImpl.cs]: cancellationToken

File: src/ResourceManagement/Compute/GalleryImpl.cs [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+        public IEnumerable<Microsoft.Azure.Management.Compute.Fluent.IGalleryImage> ListImages()
+        {
+            return this.Manager.GalleryImages.ListByGallery(this.ResourceGroupName, this.Name);
+        }
+
+        ///GENMHASH:C9EE1B32D7F72680B29C1B8F3E2216D7:CCA078D6E68AFC9EC4349D75FFE4D7AA
+        public async Task<IPagedCollection<Microsoft.Azure.Management.Compute.Fluent.IGalleryImage>> ListImagesAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            return await this.Manager.GalleryImages.ListByGalleryAsync(this.ResourceGroupName, this.Name);
+        }
```
Comment (hovsepm) [inline src/ResourceManagement/Compute/GalleryImpl.cs]: cancellationToken

File: Tests/Fluent.Tests/Compute/SharedGalleryImageTests.cs [hunk 13]
```diff
@@ -0,0 +1,337 @@
...
+                    //
+                    // Create an image in the gallery
+                    //
+                    IGalleryImage galleryImage = azure.GalleryImages.Define(galleryImageName)
+                            .WithExistingGallery(javaGallery)
+                            .WithLocation(region)
+                            .WithIdentifier("JavaSDKTeam", "JDK", "Jdk-9")
+                            .WithGeneralizedWindows()
+                            // Optionals - Start
+                            .WithUnSupportedDiskType(DiskSkuTypes.StandardLRS)
```
Comment (hovsepm) [inline Tests/Fluent.Tests/Compute/SharedGalleryImageTests.cs]: >WithUnSupportedDiskType [](start = 29, length = 23)

this look weird. there is no words `Un Supported` :) Should be just `Unsupported` i.e. `WithUnsupportedDiskType`

File: Tests/Fluent.Tests/Compute/SharedGalleryImageTests.cs [hunk 14]
```diff
@@ -0,0 +1,337 @@
...
+                            .WithNewResourceGroup(rgName)
+                            .WithDescription("java's image gallery")
+                            .Create();
+                    //
+                    // Create an image in the gallery
+                    //
+                    IGalleryImage galleryImage = azure.GalleryImages.Define(galleryImageName)
+                            .WithExistingGallery(javaGallery)
+                            .WithLocation(region)
+                            .WithIdentifier("JavaSDKTeam", "JDK", "Jdk-9")
```
Comment (hovsepm) [inline Tests/Fluent.Tests/Compute/SharedGalleryImageTests.cs]: >WithIdentifier [](start = 29, length = 14)

I think this should be plural right? `WithIdentifierS` - you seems to provide a list of Identifiers... like `WithTags`

File: Tests/Fluent.Tests/Compute/SharedGalleryImageTests.cs [hunk 15]
```diff
@@ -0,0 +1,337 @@
...
+                    //
+
+                    string versionName = "0.0.4";
+
+                    IGalleryImageVersion imageVersion = azure.GalleryImageVersions.Define(versionName)
+                            .WithExistingImage(rgName, gallery.Name, galleryImage.Name)
+                            .WithLocation(region.ToString())
+                            .WithSourceCustomImage(customImage)
+                            // Options - Start
+                            .WithAvailableInRegion(Region.USEast2)
```
Comment (hovsepm) [inline Tests/Fluent.Tests/Compute/SharedGalleryImageTests.cs]: >WithAvailableInRegion [](start = 29, length = 21)

`WithRegionAvailability` ?

File: Tests/Fluent.Tests/Compute/SharedGalleryImageTests.cs [hunk 16]
```diff
@@ -0,0 +1,337 @@
...
+                    Assert.Equal(2, imageVersion.AvailableRegions.Count);
+                    Assert.NotNull(imageVersion.ScaleTier);
+                    Assert.True(imageVersion.ScaleTier.Equals(ScaleTier.S30));
+                    Assert.False(imageVersion.IsExcludedFromLatest);
+
+                    //
+                    // Update image version
+                    //
+                    imageVersion.Update()
+                            .WithoutAvailableInRegion(Region.USEast2)
```
Comment (hovsepm) [inline Tests/Fluent.Tests/Compute/SharedGalleryImageTests.cs]: >WithoutAvailableInRegion [](start = 29, length = 24)

just IMHO - `WithoutRegionAvailability` sounds better

**Comment by hovsepm** [comment]
@anuchandy  please address review comments.

File: Tests/Fluent.Tests/Compute/SharedGalleryImageTests.cs (hunk 14 above)
Comment (anuchandy) [reply Tests/Fluent.Tests/Compute/SharedGalleryImageTests.cs]: an identifier is made of 3 parts, we are not taking multiple identifiers here.

File: Tests/Fluent.Tests/Compute/SharedGalleryImageTests.cs (hunk 13 above)
Comment (anuchandy) [reply Tests/Fluent.Tests/Compute/SharedGalleryImageTests.cs]: good catch :)

File: Tests/Fluent.Tests/Compute/SharedGalleryImageTests.cs (hunk 16 above)
Comment (anuchandy) [reply Tests/Fluent.Tests/Compute/SharedGalleryImageTests.cs]: i like `WithRegionAvailability`, `WithoutRegionAvailability` better, will update.

File: Tests/Fluent.Tests/Compute/SharedGalleryImageTests.cs (hunk 15 above)
Comment (anuchandy) [reply Tests/Fluent.Tests/Compute/SharedGalleryImageTests.cs]: thinking more about it.. we also have a wither that takes multiple regions `WithAvailableInRegions(List<Region>)`. For consistency if we replace it with `WithRegionsAvailability(List<Region>)`, not sure how readable it is.

Comment (anuchandy) [reply Tests/Fluent.Tests/Compute/SharedGalleryImageTests.cs]: ok, discussed with Asir, he recommended having overloads of `WithAvailableInRegion` that takes single region and list. Updated the PR.

File: Tests/Fluent.Tests/Compute/SharedGalleryImageTests.cs [hunk 17]
```diff
@@ -0,0 +1,337 @@
...
+                    Assert.NotNull(galleryImage.RecommendedVirtualMachineConfiguration.VCPUs.Max);
+                    Assert.Equal(25, galleryImage.RecommendedVirtualMachineConfiguration.VCPUs.Max);
+                    Assert.NotNull(galleryImage.RecommendedVirtualMachineConfiguration.Memory);
+                    Assert.NotNull(galleryImage.RecommendedVirtualMachineConfiguration.Memory.Max);
+                    Assert.Equal(3200, galleryImage.RecommendedVirtualMachineConfiguration.Memory.Max);
+                    //
+                    // Update an image in the gallery
+                    //
+                    galleryImage.Update()
+                            .WithoutUnSupportedDiskType(DiskSkuTypes.PremiumLRS)
```
[the remark this answers is not in the document]
Comment (anuchandy) [reply Tests/Fluent.Tests/Compute/SharedGalleryImageTests.cs]: thanks! corrected it now, all withers with `Unsupported` txt should be fixed now.

**Pull request merged**